crates/miden-objects/src/transaction/script_builder.rs

```diff
@@ -0,0 +1,314 @@
+        Self { assembler, libraries: Vec::new() }
+    }
+
+    /// Creates a new ScriptBuilder with a provided assembler.
+    ///
+    /// This is the recommended constructor when you need access to transaction kernel
+    /// procedures. Pass `TransactionKernel::assembler()` as the assembler parameter.
+    ///
+    /// # Arguments
+    /// * `assembler` - A pre-configured assembler (e.g., from TransactionKernel::assembler())
+    /// * `in_debug_mode` - Whether to enable debug mode
+    pub fn with_assembler(assembler: Assembler, in_debug_mode: bool) -> Self {
+        let assembler = assembler.with_debug_mode(in_debug_mode);
+        Self { assembler, libraries: Vec::new() }
+    }
+
+    // LIBRARY MANAGEMENT
+    // --------------------------------------------------------------------------------------------
+
+    /// Adds multiple libraries to the script builder.
+    ///
+    /// # Arguments
+    /// * `libraries` - Vector of compiled libraries to add to the builder
+    pub fn add_library(&mut self, libraries: Vec<Library>) -> Result<(), ScriptBuilderError> {
+        for library in libraries {
+            self.libraries.push(library);
+        }
+        Ok(())
+    }
+
+    // SCRIPT COMPILATION
+    // --------------------------------------------------------------------------------------------
+
+    /// Compiles a transaction script with the provided program code.
+    ///
+    /// The compiled script will have access to all libraries that have been added to this builder.
+    ///
+    /// # Arguments
+    /// * `program` - The transaction script source code
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - Any of the libraries cannot be loaded into the assembler
+    /// - The transaction script compilation fails
+    pub fn compile_tx_script(
+        &self,
+        tx_script: &str,
+    ) -> Result<TransactionScript, ScriptBuilderError> {
+        let mut assembler = self.assembler.clone();
+
+        // Add all libraries from the builder to the assembler
+        for lib in &self.libraries {
+            assembler = assembler.with_library(lib.clone()).map_err(|err| {
+                ScriptBuilderError::TransactionScriptLibraryError(alloc::format!(
+                    "Failed to add library: {err}"
+                ))
+            })?;
+        }
+
+        TransactionScript::compile(tx_script, assembler)
+            .map_err(ScriptBuilderError::TransactionScriptError)
+    }
+
+    /// Compiles a note script with the provided program code.
+    ///
+    /// The compiled script will have access to all libraries that have been added to this builder.
+    ///
+    /// # Arguments
+    /// * `program` - The note script source code
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - Any of the libraries cannot be loaded into the assembler
+    /// - The note script compilation fails
+    pub fn compile_note_script(&self, program: &str) -> Result<NoteScript, ScriptBuilderError> {
```

Comment:
Similar to the above comment: could we use `impl AsRef<str>` instead of `&str` here?